# Recreating the CSV datasets used in Analyzing Flight Delays and Cancellations
The following recreates the CSV data files (flights2022.csv, flights_weather2022.csv) used in the project Analyzing Flight Delays and Cancellations. Original source files are from the [Moderndive pnwflights22 Data Files](https://github.com/moderndive/pnwflights22/tree/main/data). 

Files used are in **RData format**
 - flights.rda
 - weather.rda
 - airlines.rda

I'll use the [pyreadr.read_r](https://ofajardo.github.io/pyreadr/_build/html/index.html) to load the data into a pandas dataframe and perform the necessary data analysis and sub-setting. This is my first time using this package. 

**Observations:**
- The `flights` data has a `time_hour` column which appears to combine the `year`, `month`, `day` and `hour` column information into a single field.
- The `weather` data also as a `time_hour` column which **"appears"** to combine the `year`, `month`, `day` and `hour` column information into a single field, but the hour field does not match the `time_hour` **hour** portion. The offset appears to be exactly 8 hours which would seem to indicate the UTC +8 hour offset was used. In all likelihood the `time_hour` column is the UTC value and the remaining time fields are local time. This makes a huge difference when trying to compare flight data with the weather data. 

**Weather data**
|   year |   month |   hour | time_hour           |
|-------:|--------:|-------:|:--------------------|
|   2022 |       1 |      0 | 2022-01-01 08:00:00 |
|   2022 |       1 |      1 | 2022-01-01 09:00:00 |
|   2022 |       1 |      2 | 2022-01-01 10:00:00 |
|   2022 |       1 |      3 | 2022-01-01 11:00:00 |
|   2022 |       1 |      4 | 2022-01-01 12:00:00 |
|   2022 |       1 |      5 | 2022-01-01 13:00:00 |
|   2022 |       1 |      6 | 2022-01-01 14:00:00 |
|   2022 |       1 |      7 | 2022-01-01 15:00:00 |
|   2022 |       1 |      8 | 2022-01-01 16:00:00 |
|   2022 |       1 |      9 | 2022-01-01 17:00:00 |

**Flights data**
|   year |   month |   hour | time_hour           |
|-------:|--------:|-------:|:--------------------|
|   2022 |       1 |     23 | 2022-01-01 23:00:00 |
|   2022 |       1 |     22 | 2022-01-01 22:00:00 |
|   2022 |       1 |     23 | 2022-01-01 23:00:00 |
|   2022 |       1 |     23 | 2022-01-01 23:00:00 |
|   2022 |       1 |     23 | 2022-01-01 23:00:00 |
|   2022 |       1 |     23 | 2022-01-01 23:00:00 |
|   2022 |       1 |      0 | 2022-01-01 00:00:00 |
|   2022 |       1 |     22 | 2022-01-01 22:00:00 |
|   2022 |       1 |     23 | 2022-01-01 23:00:00 |
|   2022 |       1 |     23 | 2022-01-01 23:00:00 |

In [158]:
# load our packages
import pyreadr
import pandas as pd

Inspect the RData files.

In [159]:
# examine the flights data as a list
flights_data_list = pyreadr.list_objects('flights.rda')
print(f'Flights data:\n {flights_data_list}\n')

# examine the weather data as a list
weather_data_list = pyreadr.list_objects('weather.rda')
print(f'Weather data:\n{weather_data_list}\n')

# examine the weather data as a list
airlines_data_list = pyreadr.list_objects('airlines.rda')
print(f'Airlines data:\n{airlines_data_list}')

Flights data:
 [{'object_name': 'flights', 'columns': ['year', 'month', 'day', 'dep_time', 'sched_dep_time', 'dep_delay', 'arr_time', 'sched_arr_time', 'arr_delay', 'carrier', 'flight', 'tailnum', 'origin', 'dest', 'air_time', 'distance', 'hour', 'minute', 'time_hour']}]

Weather data:
[{'object_name': 'weather', 'columns': ['origin', 'year', 'month', 'day', 'hour', 'temp', 'dewp', 'humid', 'wind_dir', 'wind_speed', 'wind_gust', 'precip', 'pressure', 'visib', 'time_hour']}]

Airlines data:
[{'object_name': 'airlines', 'columns': ['carrier', 'name']}]


In [160]:
# read the data from flights
flights_data = pyreadr.read_r('flights.rda')
print(type(flights_data))

# Print the first 5 rows of each object in the .rda file
for key, value in flights_data.items():
    print(f"Object Name: {key}")
    if isinstance(value, pd.DataFrame):
        print(value.head(5))
    else:
        print(value[:5])  # Assuming other objects are indexable like arrays
    print("\n")

# extract the dataframe into flights
flights = flights_data['flights']

# Display our flights data
display(flights.head())

<class 'collections.OrderedDict'>
Object Name: flights
   year  month  day dep_time  sched_dep_time  dep_delay arr_time  \
0  2022      1    1        1            2359        2.0      604   
1  2022      1    1        1            2250       71.0      242   
2  2022      1    1       10            2355       15.0      759   
3  2022      1    1       25            2350       35.0      606   
4  2022      1    1       35            2349       46.0      616   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             618      -14.0      UA     555  N405UA    SEA  IAH     221.0   
1             142       60.0      AS      72  N265AK    SEA  FAI     193.0   
2             730       29.0      AS     270  N274AK    SEA  ATL     261.0   
3             550       16.0      AS       7  N281AK    SEA  ORD     193.0   
4             545       31.0      UA     507  N426UA    PDX  ORD     196.0   

   distance  hour  minute           time_hour  
0    1874.0  23.0  

year  month  day dep_time  sched_dep_time  dep_delay arr_time  \
0  2022      1    1        1            2359        2.0      604   
1  2022      1    1        1            2250       71.0      242   
2  2022      1    1       10            2355       15.0      759   
3  2022      1    1       25            2350       35.0      606   
4  2022      1    1       35            2349       46.0      616   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             618      -14.0      UA     555  N405UA    SEA  IAH     221.0   
1             142       60.0      AS      72  N265AK    SEA  FAI     193.0   
2             730       29.0      AS     270  N274AK    SEA  ATL     261.0   
3             550       16.0      AS       7  N281AK    SEA  ORD     193.0   
4             545       31.0      UA     507  N426UA    PDX  ORD     196.0   

   distance  hour  minute           time_hour  
0    1874.0  23.0    59.0 2022-01-01 23:00:00  
1    1533.0  22.0    50.0 2022-01-01 22:00:00  
2    2182.0  23.0    55.0 2022-01-01 23:00:00  
3    1721.0  23.0    50.0 2022-01-01 23:00:00  
4    1739.0  23.0    49.0 2022-01-01 23:00:00

**Simplified method of loading the RData**

In [161]:
flights = pyreadr.read_r('flights.rda')['flights']
display(flights.head())
print(f'Number of columns: {len(flights.columns)}\n{flights.columns}')
print(f'Number of rows: {flights.shape[0]}')
flights.info()

year  month  day dep_time  sched_dep_time  dep_delay arr_time  \
0  2022      1    1        1            2359        2.0      604   
1  2022      1    1        1            2250       71.0      242   
2  2022      1    1       10            2355       15.0      759   
3  2022      1    1       25            2350       35.0      606   
4  2022      1    1       35            2349       46.0      616   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             618      -14.0      UA     555  N405UA    SEA  IAH     221.0   
1             142       60.0      AS      72  N265AK    SEA  FAI     193.0   
2             730       29.0      AS     270  N274AK    SEA  ATL     261.0   
3             550       16.0      AS       7  N281AK    SEA  ORD     193.0   
4             545       31.0      UA     507  N426UA    PDX  ORD     196.0   

   distance  hour  minute           time_hour  
0    1874.0  23.0    59.0 2022-01-01 23:00:00  
1    1533.0  22.0    50.0 2022-01-01 22:00:00  
2    2182.0  23.0    55.0 2022-01-01 23:00:00  
3    1721.0  23.0    50.0 2022-01-01 23:00:00  
4    1739.0  23.0    49.0 2022-01-01 23:00:00

Number of columns: 19
Index(['year', 'month', 'day', 'dep_time', 'sched_dep_time', 'dep_delay',
       'arr_time', 'sched_arr_time', 'arr_delay', 'carrier', 'flight',
       'tailnum', 'origin', 'dest', 'air_time', 'distance', 'hour', 'minute',
       'time_hour'],
      dtype='object')
Number of rows: 233385
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 233385 entries, 0 to 233384
Data columns (total 19 columns):
 #   Column          Non-Null Count   Dtype         
---  ------          --------------   -----         
 0   year            233385 non-null  int32         
 1   month           233385 non-null  int32         
 2   day             233385 non-null  int32         
 3   dep_time        228861 non-null  object        
 4   sched_dep_time  233385 non-null  int32         
 5   dep_delay       228861 non-null  float64       
 6   arr_time        228654 non-null  object        
 7   sched_arr_time  233385 non-null  int32         
 8   arr_delay       228246 non-null  float64   

In [162]:
weather = pyreadr.read_r('weather.rda')['weather']
display(weather.head())
print(f'Number of columns: {len(weather.columns)}\n{weather.columns}')
print(f'Number of rows: {weather.shape[0]}')
weather.info()

origin  year  month  day  hour  temp  dewp  humid  wind_dir  wind_speed  \
0    PDX  2022      1    1     0  32.0  21.0  63.45     120.0    11.50780   
1    PDX  2022      1    1     1  30.9  19.0  60.92     120.0    12.65858   
2    PDX  2022      1    1     2  30.9  19.0  60.92     120.0    11.50780   
3    PDX  2022      1    1     3  30.0  17.1  58.23     120.0    12.65858   
4    PDX  2022      1    1     4  28.9  17.1  60.90     120.0    11.50780   

   wind_gust  precip  pressure  visib           time_hour  
0  13.242946     0.0    1019.5   10.0 2022-01-01 08:00:00  
1  14.567241     0.0    1020.2   10.0 2022-01-01 09:00:00  
2  13.242946     0.0    1021.0   10.0 2022-01-01 10:00:00  
3  14.567241     0.0    1021.9   10.0 2022-01-01 11:00:00  
4  13.242946     0.0    1022.3   10.0 2022-01-01 12:00:00

Number of columns: 15
Index(['origin', 'year', 'month', 'day', 'hour', 'temp', 'dewp', 'humid',
       'wind_dir', 'wind_speed', 'wind_gust', 'precip', 'pressure', 'visib',
       'time_hour'],
      dtype='object')
Number of rows: 17393
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17393 entries, 0 to 17392
Data columns (total 15 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   origin      17393 non-null  object        
 1   year        17393 non-null  int32         
 2   month       17393 non-null  int32         
 3   day         17393 non-null  int32         
 4   hour        17393 non-null  int32         
 5   temp        17393 non-null  float64       
 6   dewp        17393 non-null  float64       
 7   humid       17393 non-null  float64       
 8   wind_dir    16810 non-null  float64       
 9   wind_speed  17370 non-null  float64       
 10  wind_gust   17370 non-null  float64       
 11  precip      17393 non-nu

The flights data has 233385 rows but the dataset we were working with in the project only has 111376. The dataset from the project focused on the first 6 months of the year. 

In [163]:
# only use the first six months of data
flights = flights.query('month <= 6')
flights.shape

(111376, 19)

Our flights dataset also had an airline column and will grap that data from the `airlines.rda` file.

In [164]:
# read in the airlines data.
airlines = pyreadr.read_r('airlines.rda')['airlines']
display(airlines.head())
airlines.info()

carrier                    name
0      AA  American Airlines Inc.
1      AS    Alaska Airlines Inc.
2      B6         JetBlue Airways
3      DL    Delta Air Lines Inc.
4      F9  Frontier Airlines Inc.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12 entries, 0 to 11
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   carrier  12 non-null     object
 1   name     12 non-null     object
dtypes: object(2)
memory usage: 324.0+ bytes


In [165]:
# rename the columns to match what we need in flights.
airlines.columns = ['carrier', 'airline']
airlines.head()

carrier                 airline
0      AA  American Airlines Inc.
1      AS    Alaska Airlines Inc.
2      B6         JetBlue Airways
3      DL    Delta Air Lines Inc.
4      F9  Frontier Airlines Inc.

Add the Airlines name to our `flights` dataframe.

In [166]:
flights = flights.merge(airlines, on='carrier', how='inner')

In [167]:
flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 111376 entries, 0 to 111375
Data columns (total 20 columns):
 #   Column          Non-Null Count   Dtype         
---  ------          --------------   -----         
 0   year            111376 non-null  int32         
 1   month           111376 non-null  int32         
 2   day             111376 non-null  int32         
 3   dep_time        108931 non-null  object        
 4   sched_dep_time  111376 non-null  int32         
 5   dep_delay       108931 non-null  float64       
 6   arr_time        108834 non-null  object        
 7   sched_arr_time  111376 non-null  int32         
 8   arr_delay       108697 non-null  float64       
 9   carrier         111376 non-null  object        
 10  flight          111376 non-null  int32         
 11  tailnum         111247 non-null  object        
 12  origin          111376 non-null  object        
 13  dest            111376 non-null  object        
 14  air_time        108697 non-null  flo

Ok, that closely matches the original flights2022.csv data and can save it off. 

In [168]:
# save off the dataframe to csv.
flights.to_csv('output/flights2022_recreated.csv', index=False)

Now onto the weather dataframe.

The original weather dataframe has 111006 rows and 29 columns. Our weather data only has 17393 rows.

The original weather dataframe has the weather data assigned to each flight, so I'll do the same. 

To be accurate I'll need to match against `origin`, `year`, `month`, `day` and `hour`.

**Additional observations:**
- The original dataframe had a `route` column but no `time_hour` column so will need to add that.
- Our Rdata files have the `time_hour` column but no `route` column.

Based on the column order in `flights2022.csv` and `flights_weather2022.csv` I surmise that:
- the flights Rdata above was saved to the csv as above and
- then a `route` column was added to the dataframe and 
- then the weather data was merged.
- the `time_hour` data column was dropped and
- then the flights_weather2022.csv file saved. 

Let's see!

Add the `route` column to flights

In [169]:
flights['route'] = flights.origin + '-' + flights.dest
flights[['origin', 'dest', 'route',]].head()

origin dest    route
0    SEA  IAH  SEA-IAH
1    PDX  ORD  PDX-ORD
2    SEA  IAD  SEA-IAD
3    SEA  ORD  SEA-ORD
4    SEA  DEN  SEA-DEN

In [170]:
# common columns between our two data sources
common_columns = list(flights.columns.intersection(weather.columns))
print(common_columns)

['year', 'month', 'day', 'origin', 'hour', 'time_hour']


We want to pull in the data from the weather data into the flights dataframe. The weather data is for all of 2022 and appears to be recorded at 1 hour intervals as it does not contain a `minute` column. Matching on the `origin` and `time_hour` column should simplify things instead of using the full list of columns `['origin', 'year', 'month', 'day','hour', 'time_hour]`, but this would leave us with suffixes to remove.  

**Reviewing our common columns.**

In [171]:
flights[['origin', 'year', 'month', 'day','hour', 'minute', 'time_hour' ]].head()

origin  year  month  day  hour  minute           time_hour
0    SEA  2022      1    1  23.0    59.0 2022-01-01 23:00:00
1    PDX  2022      1    1  23.0    49.0 2022-01-01 23:00:00
2    SEA  2022      1    1  23.0     5.0 2022-01-01 23:00:00
3    SEA  2022      1    1  23.0    55.0 2022-01-01 23:00:00
4    SEA  2022      1    1   5.0     5.0 2022-01-01 05:00:00

In [172]:
weather[['origin', 'year', 'month', 'day','hour', 'time_hour']].head()

origin  year  month  day  hour           time_hour
0    PDX  2022      1    1     0 2022-01-01 08:00:00
1    PDX  2022      1    1     1 2022-01-01 09:00:00
2    PDX  2022      1    1     2 2022-01-01 10:00:00
3    PDX  2022      1    1     3 2022-01-01 11:00:00
4    PDX  2022      1    1     4 2022-01-01 12:00:00

**Note:** The `time_hour` column in weather doesn't align with the `hour` column. I'll perform the merge and see what we end up with. 

Based on the original `flights_weather2022.csv` data the number of records is less than the `flights2022.csv` data set, so an inner join was likely performed. 

In [173]:
flights_weather = flights.merge(weather, how='inner', on=['origin', 'time_hour'], suffixes=(None, '_weather'))
display(flights_weather.head())
print(flights_weather.shape)

year  month  day dep_time  sched_dep_time  dep_delay arr_time  \
0  2022      1    1        1            2359        2.0      604   
1  2022      1    1      142            2305      157.0      900   
2  2022      1    1      255            2355      180.0      831   
3  2022      1    1       10            2355       15.0      759   
4  2022      1    1       25            2350       35.0      606   

   sched_arr_time  arr_delay carrier  ...  hour_weather  temp  dewp  humid  \
0             618      -14.0      UA  ...            15  26.1  17.1  68.34   
1             657      123.0      UA  ...            15  26.1  17.1  68.34   
2             549      162.0      UA  ...            15  26.1  17.1  68.34   
3             730       29.0      AS  ...            15  26.1  17.1  68.34   
4             550       16.0      AS  ...            15  26.1  17.1  68.34   

   wind_dir  wind_speed  wind_gust  precip pressure visib  
0     140.0     9.20624  10.594357     0.0   1025.6  10.0  
1     140.0     9.20624  10.594357     0.0   1025.6  10.0  
2     140.0     9.20624  10.594357     0.0   1025.6  10.0  
3     140.0     9.20624  10.594357     0.0   1025.6  10.0  
4     140.0     9.20624  10.594357     0.0   1025.6  10.0  

[5 rows x 34 columns]

(110875, 34)


We end up with 131 records less than expected, so `time_hour` was not what was used to perform the merge, and we also end up with additional suffixed columns we don't need. If we perform a merge on all the `common_columns` we end up with 0 rows matching. This will be due to the discrepancy between the `hour` and `time_hour` columns in the weather data. 


In [174]:
flights_weather = flights.merge(weather, how='inner', on=common_columns)
display(flights_weather.head())
print(flights_weather.columns)

Empty DataFrame
Columns: [year, month, day, dep_time, sched_dep_time, dep_delay, arr_time, sched_arr_time, arr_delay, carrier, flight, tailnum, origin, dest, air_time, distance, hour, minute, time_hour, airline, route, temp, dewp, humid, wind_dir, wind_speed, wind_gust, precip, pressure, visib]
Index: []

[0 rows x 30 columns]

Index(['year', 'month', 'day', 'dep_time', 'sched_dep_time', 'dep_delay',
       'arr_time', 'sched_arr_time', 'arr_delay', 'carrier', 'flight',
       'tailnum', 'origin', 'dest', 'air_time', 'distance', 'hour', 'minute',
       'time_hour', 'airline', 'route', 'temp', 'dewp', 'humid', 'wind_dir',
       'wind_speed', 'wind_gust', 'precip', 'pressure', 'visib'],
      dtype='object')


Removing the `time_hour` column and performing the merge again.

In [175]:
common_columns.remove('time_hour')
print(common_columns)

flights_weather = flights.merge(weather, how='inner', on=common_columns, suffixes=('_flights', '_weather'))
display(flights_weather.head())
print(flights_weather.columns)
print(flights_weather.shape)

['year', 'month', 'day', 'origin', 'hour']


year  month  day dep_time  sched_dep_time  dep_delay arr_time  \
0  2022      1    1        1            2359        2.0      604   
1  2022      1    1      142            2305      157.0      900   
2  2022      1    1      255            2355      180.0      831   
3  2022      1    1       10            2355       15.0      759   
4  2022      1    1       25            2350       35.0      606   

   sched_arr_time  arr_delay carrier  ...  temp  dewp  humid wind_dir  \
0             618      -14.0      UA  ...  33.0  23.0  66.06    160.0   
1             657      123.0      UA  ...  33.0  23.0  66.06    160.0   
2             549      162.0      UA  ...  33.0  23.0  66.06    160.0   
3             730       29.0      AS  ...  33.0  23.0  66.06    160.0   
4             550       16.0      AS  ...  33.0  23.0  66.06    160.0   

   wind_speed  wind_gust  precip  pressure visib   time_hour_weather  
0     8.05546   9.270062     0.0    1022.9  10.0 2022-01-02 07:00:00  
1     8.05546   9.270062     0.0    1022.9  10.0 2022-01-02 07:00:00  
2     8.05546   9.270062     0.0    1022.9  10.0 2022-01-02 07:00:00  
3     8.05546   9.270062     0.0    1022.9  10.0 2022-01-02 07:00:00  
4     8.05546   9.270062     0.0    1022.9  10.0 2022-01-02 07:00:00  

[5 rows x 31 columns]

Index(['year', 'month', 'day', 'dep_time', 'sched_dep_time', 'dep_delay',
       'arr_time', 'sched_arr_time', 'arr_delay', 'carrier', 'flight',
       'tailnum', 'origin', 'dest', 'air_time', 'distance', 'hour', 'minute',
       'time_hour_flights', 'airline', 'route', 'temp', 'dewp', 'humid',
       'wind_dir', 'wind_speed', 'wind_gust', 'precip', 'pressure', 'visib',
       'time_hour_weather'],
      dtype='object')
(111006, 31)


We've got the right number of rows, and we can check that the `time_hour` from weather is what is causing an issue with the previous merges. 

In [176]:
flights_weather[['month', 'day', 'hour', 'time_hour_weather', 'time_hour_flights']].head()

month  day  hour   time_hour_weather   time_hour_flights
0      1    1  23.0 2022-01-02 07:00:00 2022-01-01 23:00:00
1      1    1  23.0 2022-01-02 07:00:00 2022-01-01 23:00:00
2      1    1  23.0 2022-01-02 07:00:00 2022-01-01 23:00:00
3      1    1  23.0 2022-01-02 07:00:00 2022-01-01 23:00:00
4      1    1  23.0 2022-01-02 07:00:00 2022-01-01 23:00:00

We can drop the `time_hour` columns as we don't have them in the `flights_weather2022.csv` data.

In [177]:
flights_weather.drop(columns=['time_hour_flights', 'time_hour_weather'], inplace=True)
print(flights_weather.columns)
print(flights_weather.shape)

Index(['year', 'month', 'day', 'dep_time', 'sched_dep_time', 'dep_delay',
       'arr_time', 'sched_arr_time', 'arr_delay', 'carrier', 'flight',
       'tailnum', 'origin', 'dest', 'air_time', 'distance', 'hour', 'minute',
       'airline', 'route', 'temp', 'dewp', 'humid', 'wind_dir', 'wind_speed',
       'wind_gust', 'precip', 'pressure', 'visib'],
      dtype='object')
(111006, 29)


There we go, now we can save our data to CSV.

In [178]:
flights_weather.to_csv('output/flights_weather2022_recreated.csv', index=False)